File: go/internal/ui/accordion_item.go
package ui

import (
	"fmt"
	"strings"
)

// AccordionState represents the possible states of an accordion item
type AccordionState string

const (
	// AccordionStateOpened represents an opened accordion item
	AccordionStateOpened AccordionState = "opened"
	// AccordionStateClosed represents a closed accordion item
	AccordionStateClosed AccordionState = "closed"
)

// AccordionItem represents an accordion item component
type AccordionItem struct {
	Title         string
	State         AccordionState
	ButtonOptions map[string]interface{}
	Content       string
}

// NewAccordionItem creates a new accordion item with the given parameters
func NewAccordionItem(title string, state AccordionState, buttonOptions map[string]interface{}, content string) *AccordionItem {
	if buttonOptions == nil {
		buttonOptions = make(map[string]interface{})
	}

	// Set default button options if not provided
	if _, ok := buttonOptions["aria-controls"]; !ok {
		buttonOptions["aria-controls"] = "accordion-item"
	}

	return &AccordionItem{
		Title:         title,
		State:         state,
		ButtonOptions: buttonOptions,
		Content:       content,
	}
}

// Icon returns the appropriate icon name based on the accordion state
func (a *AccordionItem) Icon() string {
	if a.State == AccordionStateOpened {
		return "chevron-down"
	}
	return "chevron-right"
}

// BodyClass returns the CSS classes for the accordion body
func (a *AccordionItem) BodyClass() map[string]interface{} {
	if a.State == AccordionStateOpened {
		return map[string]interface{}{
			"class": "show",
		}
	}
	return map[string]interface{}{}
}

// IsExpanded returns whether the accordion item is expanded
func (a *AccordionItem) IsExpanded() bool {
	return a.State == AccordionStateOpened
}

// Render generates the HTML for the accordion item
func (a *AccordionItem) Render() string {
	// Build button attributes
	buttonAttrs := make(map[string]interface{})
	for k, v := range a.ButtonOptions {
		buttonAttrs[k] = v
	}
	buttonAttrs["aria-expanded"] = a.IsExpanded()

	// Build body attributes
	bodyAttrs := a.BodyClass()
	bodyAttrs["class"] = fmt.Sprintf("accordion-item gl-mt-3 gl-text-base collapse %s",
		strings.Join(bodyAttrs["class"].([]string), " "))

	return fmt.Sprintf(`<div class="gl-accordion-item">
	<h3 class="gl-accordion-item-header">
		<button type="button" %s>
			<span class="js-chevron-icon">%s</span>
			%s
		</button>
	</h3>
	<div %s>
		%s
	</div>
</div>`,
		formatAttributes(buttonAttrs),
		a.Icon(),
		a.Title,
		formatAttributes(bodyAttrs),
		a.Content)
}

File: rust_app/src/components/pajamas/accordion_item.rs
use super::button::ButtonComponent;
use std::collections::HashMap;

/// State options for the accordion item
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AccordionItemState {
    Opened,
    Closed,
}

impl AccordionItemState {
    fn as_str(&self) -> &'static str {
        match self {
            AccordionItemState::Opened => "opened",
            AccordionItemState::Closed => "closed",
        }
    }
}

/// Renders a Pajamas compliant accordion item element
#[derive(Debug, Clone)]
pub struct AccordionItemComponent {
    /// The title of the accordion item
    title: Option<String>,
    /// The state of the accordion item
    state: AccordionItemState,
    /// Options for the button
    button_options: HashMap<String, String>,
    /// The content of the accordion item
    content: Option<String>,
}

impl AccordionItemComponent {
    /// Create a new accordion item component
    pub fn new(
        title: Option<String>,
        state: AccordionItemState,
        button_options: Option<HashMap<String, String>>,
    ) -> Self {
        Self {
            title,
            state,
            button_options: button_options.unwrap_or_default(),
            content: None,
        }
    }

    /// Set the content of the accordion item
    pub fn with_content(mut self, content: String) -> Self {
        self.content = Some(content);
        self
    }

    /// Get the icon for the accordion item
    fn icon(&self) -> &'static str {
        match self.state {
            AccordionItemState::Opened => "chevron-down",
            AccordionItemState::Closed => "chevron-right",
        }
    }

    /// Get the body class for the accordion item
    fn body_class(&self) -> HashMap<String, String> {
        let mut classes = HashMap::new();
        if self.state == AccordionItemState::Opened {
            classes.insert("class".to_string(), "show".to_string());
        }
        classes
    }

    /// Check if the accordion item is expanded
    fn expanded(&self) -> bool {
        self.state == AccordionItemState::Opened
    }

    /// Render the accordion item component
    pub fn render(&self) -> String {
        let mut html = String::new();

        // Start accordion item
        html.push_str(r#"<div class="gl-accordion-item">"#);

        // Render header
        html.push_str(r#"<h3 class="gl-accordion-item-header">"#);

        // Render button
        if let Some(title) = &self.title {
            let mut button = ButtonComponent::new();
            button.variant = super::button::ButtonVariant::Link;
            button.icon = Some(self.icon().to_string());
            button.icon_classes = Some("js-chevron-icon".to_string());
            button.content = Some(title.clone());

            // Add button options
            button.button_options = self.button_options.clone();
            button
                .button_options
                .insert("aria-controls".to_string(), "accordion-item".to_string());
            button
                .button_options
                .insert("aria-expanded".to_string(), self.expanded().to_string());

            html.push_str(&button.render());
        }

        html.push_str("</h3>");

        // Render body
        let body_classes = self.body_class();
        let body_class_str = body_classes
            .get("class")
            .map(|c| format!(" {}", c))
            .unwrap_or_default();

        html.push_str(&format!(
            r#"<div class="accordion-item gl-mt-3 gl-text-base collapse{}">"#,
            body_class_str
        ));

        // Render content
        if let Some(content) = &self.content {
            html.push_str(content);
        }

        html.push_str("</div></div>");
        html
    }
}
